### PR TITLE
Add type dropdown to edit expense options

### PR DESCRIPTION
**Pull request by jeffin143**
Handling https://github.com/opencollective/opencollective/issues/2486

Screenshot from 2019-10-18 14-41-52

Edit option, drop down 

Screenshot from 2019-10-18 14-44-10

### CODE REVIEW

**Comment by jeffin143** [comment]
>     * when editing, similar to the "New expense form", the "Type" should be mandatory, it should not be 
possible to save with a blank one
> 
Done, Added a validate function for expense 

>     * when editing an expense with a set "Type", the proper "Type" is not selected in the select input
> 
>     * not sure if it's the previous issue, but when editing again just after a first edit, the new "Type" is not selected in the select input

@znarf I am not sure, why this is buggy, tried debugging it a lot, but couldn't get the flow, Could you take a look whenever you are free.

**Comment by znarf** [comment]
@jeffin143 marking the PR as blocked, let's make sure to remove the tag when the situation change

**Comment by jeffin143** [comment]
> Testing again today:
> 
>     1. When editing the "Type" and submitting the form, the "Type" is not sent to the API. For that reason it's not updated.
> 

https://github.com/opencollective/opencollective-frontend/blob/d4e01f66e06dc0b568d8ac4ff5fe584168c2ef28/components/expenses/Expense.js#L498-L510,  following is where the expense is mutated, so would adding type field here work ? if not could you point me to the location where the error is ?

**Comment by jeffin143** [comment]
@znarf , Solved both the issues :)

**Comment by jeffin143** [comment]
> Please manually test your changes before asking for a review.

@znarf  Sorry about this, I will make sure I do it next time onward.
> 
> Maybe it's because the blank type is `UNCLASSIFIED` in the DB?

Yes I have removed it, 

Here are the images
Screenshot from 2019-11-12 21-24-40
Screenshot from 2019-11-12 21-24-28

File: components/expenses/ExpenseDetails.js [hunk 1]
```diff
@@ -86,6 +88,14 @@ class ExpenseDetails extends React.Component {
     const categoriesOptions = categories(expense.collective.slug).map(category => {
       return { [category]: category };
     });
+    const expenseTypes = { DEFAULT: '', INVOICE: 'INVOICE', RECEIPT: 'RECEIPT' };
+    // Delete default type if expense is receipt or invoice.
+    if (expense.type === 'RECEIPT' || expense.type === 'INVOICE') {
+      delete expenseTypes['DEFAULT'];
+    }
+    const expenseTypesoption = Object.entries(expenseTypes).map(([key, value]) => {
```
Comment (znarf) [inline components/expenses/ExpenseDetails.js]: `expenseTypesoption` is not camelCase, it should be `expenseTypesOption`

Comment (znarf) [reply components/expenses/ExpenseDetails.js]: And while we're at it, it should be plural: `expenseTypesOptions`

File: components/expenses/ExpenseDetails.js [hunk 2]
```diff
@@ -66,6 +70,10 @@ class ExpenseDetails extends React.Component {
       ...this.state.expense,
       [attr]: value,
     };
+    if (expense['type'] !== 'RECEIPT' && expense['type'] !== 'INVOICE') {
+      console.error(this.props.intl.formatMessage(this.messages['expenseTypeMissing']));
```
Comment (znarf) [inline components/expenses/ExpenseDetails.js]: Writing an internationalized message to the console is not something do:
- if it's for developers, they don't need internationalization
- if it's for the user, they will not see it

Comment (jeffin143) [reply components/expenses/ExpenseDetails.js]: Removed

**Comment by jeffin143** [comment]
> The user needs to see the message somehow, logging to the browser console is not useful.

isn't we handling that in separate issue ? -> https://github.com/opencollective/opencollective-frontend/pull/2865#discussion_r339554610

**Comment by znarf** [comment]
@jeffin143 No, I don't see the relationship. Here, the user just need some feedback on why the form can't be saved.

**Comment by jeffin143** [comment]
> @jeffin143 No, I don't see the relationship. Here, the user just need some feedback on why the form can't be saved.

would alert box work ?

**Comment by Betree** [comment]
@jeffin143 on the "Create expense" form we show the error as a simple red text:



Doing the same on the edit form with a `<Span color="red.500">{error}</Span>` would be nice

image

**Comment by jeffin143** [comment]
>@jeffin143 on the "Create expense" form we show the error as a simple red text:

Screenshot from 2019-11-13 22-16-32
Screenshot from 2019-11-13 22-16-25

**Comment by jeffin143** [comment]
Ran `npm run prettier:write` to solve lint issue, and that edit many files, not sure, should I revert the changes or not ?

**Comment by znarf** [comment]
@jeffin143 you should `npm install` and run `npm run prettier:write` again

**Comment by jeffin143** [comment]
> @jeffin143 you should `npm install` and run `npm run prettier:write` again. You're on the older version of prettier.

Thanks you so much :)

**Comment by jeffin143** [comment]
> Tested the last version, works great!

This PR fix took longer than I expected to

**Pull request merged**